src/GeoModel.js

```diff
@@ -1,95 +1,290 @@
 import BaseModel from './BaseModel'
 import Promise from 'bluebird'
 
-import traverse from 'traverse'
 import _ from "underscore"
+import rewind from 'geojson-rewind'
+import turf from '@turf/meta'
+import turfCentroid from '@turf/centroid'
+turf.centroid = turfCentroid
 
+/** Cleans up polygon coordinates.
+ * For each linear ring that defines the polygon, it removes repeated
+ * coordinates, ensures ring is closed. It discards rings with less than
+ * four points, and if the outer ring is discarded the polygon becomes
+ * undefined.
+ * @private
+ */
 function cleanPolygon(coordinates) {
-  return _.uniq(coordinates, false, (i) => { return i[0]+"-"+i[1] }).concat([coordinates[0]])
-  
+  function near(v1, v2) { return Math.abs(v1 - v2) < 0.000001 } // about 10 cm
+  function same(c1, c2) { return near(c1[0], c2[0]) && near(c1[1], c2[1]) }
+
+  let rslt = []
+  coordinates.forEach((ring, index) => {
+    ring = ring.reduce((acc, val, idx) => {
+      if ((idx === 0) || !same(acc[acc.length - 1], val)) acc.push(val)
+      return acc
+    }, [])
+    if (same(ring[0], ring[ring.length - 1])) ring.pop()
+    ring.push(ring[0])
+    if (ring.length < 4) ring = undefined
+    if (ring || (index === 0)) rslt.push(ring)
+  })
+  return rslt[0] && rslt
 }
 
-function extractGeometry(obj) {
-  return traverse(obj).reduce(function(acc, x) {
-    if(this.key == 'geometry') {
-      if (x.type == 'Polygon') {
-        // Filter illegal repeats of coordinates, re-add end point
-        x.coordinates[0] = cleanPolygon(x.coordinates[0])
-      }
-      if (x.type == 'MultiPolygon') {
-        // Filter illegal repeats of coordinates, re-add end point
-        x.coordinates = _.map(x.coordinates, (coords) => {
-          coords[0] = cleanPolygon(coords[0])
-          return coords
-        })
-      }
-      acc.push(x)
-    };
-    return acc;
-  }, [])
+function cleanGeometryCoordinates(parts) {
+  let cleaned = []
+  parts.Polygon.forEach((coordinates) => {
+    coordinates = cleanPolygon(coordinates)
+    if (coordinates) {
+      let obj = rewind({type: 'Polygon', coordinates})
```

Comment:
just a question here: are you using rewind as a kind of normalization, or is this about better rendering of geojson on the client? Or both?